File: 2026/Labs/NeuralVision/tests/verify_browser.py
#!/usr/bin/env python3
"""Compare browser computations with independent NumPy and held-out PyTorch outputs."""
import json
from pathlib import Path
import subprocess
import sys
import tempfile
import numpy as np

root=Path(__file__).resolve().parents[1]
sys.path.insert(0,str(root/'ws/src/neural_vision_common'))
from neural_vision_common.inference import Network
models=root/'ws/src/neural_vision_common/models'
def native(x):
    if isinstance(x,np.ndarray):return x.tolist()
    if isinstance(x,dict):return {k:native(v) for k,v in x.items()}
    if isinstance(x,list):return [native(v) for v in x]
    return x
cases=[]
for name in ('mlp','cnn','transformer'):
    net=Network(name,models)
    with np.load(models/f'{name}_reference.npz',allow_pickle=False) as refs:
        images=refs['inputs']; logits=refs['logits']
        for i,image in enumerate(images):
            result=net.forward(image)
            # All intermediate values for three inputs; predictions for all forty.
            if i>=3:result.pop('detail')
            cases.append(dict(model=name,pixels=image.reshape(-1).tolist(),options={},result=native(result),torch=logits[i].tolist()))
    variants={'mlp':[{'disabled_neuron':0},{'disabled_neuron':150},{'disabled_neuron':299}],
              'cnn':[{'negative_slope':.01},{'negative_slope':.1},{'negative_slope':.3}],
              'transformer':[{'use_positions':False}]}[name]
    for options in variants:
        cases.append(dict(model=name,pixels=images[0].reshape(-1).tolist(),options=options,result=native(net.forward(images[0],**options))))
    for image in [np.zeros((28,28),dtype=np.float32),np.ones((28,28),dtype=np.float32)]:
        cases.append(dict(model=name,pixels=image.reshape(-1).tolist(),options={},result=native(net.forward(image))))
with tempfile.TemporaryDirectory(prefix='neural-browser-') as tmp:
    tmp=Path(tmp)
    (tmp/'cases.json').write_text(json.dumps(cases,separators=(',',':')))
    for name in ('mlp','cnn','transformer'):
        with np.load(models/f'{name}.npz',allow_pickle=False) as data:
            (tmp/f'{name}.json').write_text(json.dumps({key:data[key].tolist() for key in data.files},separators=(',',':')))
    subprocess.run(['node',str(root/'tests/verify_browser.mjs'),str(tmp)],check=True)

File: 2026/Labs/NeuralVision/tests/verify_browser.mjs
import assert from 'node:assert/strict';
import fs from 'node:fs';
import path from 'node:path';
import {forward} from '../ws/src/neural_vision_common/web/browser-inference.js';
import {layersFor,inspectNode} from '../ws/src/neural_vision_common/web/network-data.js';
import {checkInspector} from './inspector_checks.mjs';
import {inputStatus} from '../ws/src/neural_vision_common/web/input-status.js';
const directory=process.argv[2], read=name=>JSON.parse(fs.readFileSync(path.join(directory,name+'.json'),'utf8'));
const weights=Object.fromEntries(['mlp','cnn','transformer'].map(n=>[n,read(n)]));
let compared=0,maxError=0,inspected=0;const fullyInspected=new Set();
function compare(actual,expected,where){
  if(typeof expected==='number'){
    const error=Math.abs(actual-expected);maxError=Math.max(maxError,error);compared++;
    assert(error<3e-5+3e-5*Math.abs(expected),`${where}: ${actual} vs ${expected}`);
  }else if(Array.isArray(expected)){
    assert.equal(actual.length,expected.length,where);
    expected.forEach((e,i)=>compare(actual[i],e,`${where}[${i}]`));
  }else if(expected&&typeof expected==='object'){
    for(const key of Object.keys(expected))compare(actual[key],expected[key],where+'.'+key);
  }else assert.equal(actual,expected,where);
}
const cases=read('cases');
for(const [i,c] of cases.entries()){
  const pixels=c.pixels.map(v=>v*255), w=weights[c.model], result=forward(c.model,pixels,w,c.options);
  compare(result,c.result,`${c.model}/${i}`);
  if(c.torch)compare(result.logits,c.torch,`${c.model}/${i}/pytorch`);
  if(c.result.detail){
    const layers=layersFor(c.model,c.pixels,result),state={model:c.model,image:c.pixels,result,weights:w,layers,head:1,dimension:3};
    const exhaustive=!fullyInspected.has(c.model);
    for(const layer of layers) {
      const channels=exhaustive?Array.from({length:layer.channels},(_,i)=>i):[0,layer.channels-1];
      const units=exhaustive?Array.from({length:layer.rows*layer.cols},(_,i)=>i):[0,Math.floor(layer.rows*layer.cols/2),layer.rows*layer.cols-1];
      for(const ch of channels)for(const unit of units){
        const record=inspectNode(state,layer,ch,unit);
        assert(Number.isFinite(record.value));
        checkInspector(state,layer,ch,unit,record,compare);
        inspected++;
      }
    }
    fullyInspected.add(c.model);
  }
}
for(const bad of [[],Array(784).fill(NaN),Array(784).fill(-1),Array(784).fill(256)])assert.throws(()=>forward('mlp',bad,weights.mlp));
const evaluated=Array(784).fill(0),draft=evaluated.slice();draft[100]=255;
assert.equal(inputStatus(evaluated,evaluated,true).kind,'current');
assert.equal(inputStatus(draft,evaluated,true).kind,'changed');
assert.equal(inputStatus(draft,evaluated,true,true).kind,'pending');
assert.equal(inputStatus(draft,draft,false).kind,'pending');
assert.equal(inputStatus(draft,draft,true).kind,'current');
console.log(`PASS ${cases.length} browser cases; ${compared.toLocaleString()} numeric comparisons; ${inspected} inspected units; max absolute error ${maxError.toExponential(3)}`);
